### DISCUSSION TITLE
How can you link a Rust library with a C app into one WASM file?

### DISCUSSION THREAD

**Discussion by curiousdannii** [emscripten-core/emscripten#17294] [Q&A] [answered]
I have a project that links a C library with a C app using Emscripten. I would like to replace the C library with a library written in Rust. How can this be accomplished?

I know that there are two relevant WASM targets in Rust, the normal wasm32-unknown-unknown one and the wasm32-unknown-emscripten one. I'm guessing the latter target is necessary for it to work with the Emscripten compiler? But how do I tell Emscripten to use a .wasm library rather than a .a library?

There may be other ways of linking Rust and C code together into a .wasm but I think I do need to stick with Emscripten as I require its Asyncify feature (and in the future will use its WASMFS system too.)

**Comment by kripken** [comment]
Given you need only pure computation from the Rust side, IIUC (no `wasm-bindgen`) then it would be best to tell Rust to emit a wasm object file. Then that can just be passed to emcc as another normal linker input (emcc accepts `.o` files as inputs). Using the `wasm32-unknown-emscripten` triple is probably safest while doing so.

I'm not sure how to get Rust to emit a `.o` file, though. But if there's a simple way to make it emit a `.a` file (a static library) containing `.o` files (and not, say, LLVM IR) that would be good as well.

A risk, however, is if Rust uses libc code, and the Rust compiler's idea of what that ABI is differs from emcc. I think Rust has that hardcoded in some form, and there is no simple way to make it use current emcc's version. In Zig for example you can tell it to use the emscripten sysroot, but I don't know if there's a simple way to do that in Rust.

cc @tlively who knows this the best

**Reply by tlively** [reply, marked as answer]
To produce a Rust .a file, set the `crate-type` field to `staticlib` in your cargo.toml: https://doc.rust-lang.org/cargo/reference/cargo-targets.html#the-crate-type-field

**Reply by curiousdannii** [reply]
Well that sounds optimistically simple, so I'll give it a try once I've rewritten the library in rust, and then report back.

I will need to use libc just for stdio, but I'm sure there's a way to make it work with Emscripten's libc.

**Reply by curiousdannii** [reply]
So just compiling a staticlib to the `wasm32-unknown-emscripten` target works! Though Rust may require the linker language to be set to C++.
